Add Map support to arrow-avro

jecsand838 opened the pull request:
# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/issues/4886

Related to https://github.com/apache/arrow-rs/pull/6965

# Rationale for this change
 
Avro supports maps as a core data type, but previously arrow-avro lacked any decoding logic to handle them. As a result, any Avro file containing map fields would fail to parse correctly within the Arrow ecosystem. This PR addresses this gap by:

1. Adding explicit Map decoding: It introduces a new` Codec::Map` variant and corresponding `Decoder::Map` logic that reads map blocks in Avro format and constructs an Arrow MapArray. This preserves both the keys and values in a schema-compatible manner, enabling accurate ingestion of map-encoded data.

Overall, these changes expand Arrow’s Avro reader capabilities, allowing users to work with map-encoded data in a standardized Arrow format.

# What changes are included in this PR?

This PR adds full Map support in the arrow-avro crate. Key changes include:

**1. arrow-avro/src/codec.rs:**

* Added a `Codec::Map(Arc<AvroDataType>)` variant
* Logic to construct an Arrow MapArray, including a struct of { key, value } with typed value decoding.
* Added a new function to `AvroDataType`.

**2. arrow-avro/src/reader/record.rs:**

* Introduced the Map decoding path which leverages blockwise reads of Avro map data.
* Created helpers read_map_blocks and `read_blockwise_items` to handle block sequences of key value pairs until a terminating block count of zero.
* Implemented decoder unit tests for Map types.

# Are there any user-facing changes?

N/A

klion26 commented on arrow-avro/src/codec.rs:
```diff
@@ -155,6 +168,22 @@ impl Codec {
             Self::Struct(f) => DataType::Struct(f.iter().map(|x| x.field()).collect()),
+            Self::Map(value_type) => {
+                let val_dt = value_type.codec.data_type();
+                let val_field = Field::new("value", val_dt, true)
```
Does the `nullability` set to `value_type.nullability.is_some()`

klion26 replied:
Read the [spec](https://avro.apache.org/docs/1.11.1/specification/#schema-declaration), did not find that if the value for `Map` can be null or not, but asked Deepseek, it says that the value can be null if the schema supports it(use `Unions` such as `["null", "string"]`), else it can't be null.

jecsand838 replied:
@klion26 

I was attempting to follow the same pattern used in the other types such as `ComplexType::Record(r)`, `ComplexType::Array(a)`, and `ComplexType::Fixed(f)`.

It seemed from reading the code that nullability would only be set in the `Schema::Union(f)` branch of `make_data_type`.

jecsand838 replied:
@klion26  That's a really good call out. I think it does make sense to add that in here.

klion26 commented on arrow-avro/src/codec.rs:
```diff
@@ -284,9 +313,14 @@ fn make_data_type<'a>(
+            ComplexType::Map(m) => {
+                let val = make_data_type(&m.values, namespace, resolver)?;
+                Ok(AvroDataType {
+                    nullability: None,
```
Do we need to set the `nullability` to `val.nullability`?

jecsand838 replied:
@klion26 

I was attempting to follow the same pattern used in the other types such as `ComplexType::Record(r)`, `ComplexType::Array(a)`, and `ComplexType::Fixed(f)`.

It seemed from reading the code that nullability would only be set in the `Schema::Union(f)` branch of `make_data_type`.

klion26 commented on arrow-avro/src/reader/record.rs:
```diff
@@ -267,10 +305,83 @@ impl Decoder {
+) -> Result<usize, ArrowError> {
+    let mut total = 0usize;
+    loop {
+        let blk = buf.get_long()?;
```
It seems `blk` is the number of items needed to be read. Is there any case where `blk` will be negative?

jecsand838 replied:
@klion26 

`blk` is the block count and there are cases where `blk` will be negative. 

A negative `blk` is expected by the Avro spec for block-encoded arrays & maps, indicating that the count is `-blk` items. Avro decoders usually handle this by reading the size marker and then proceeding to decode `|blk|` entries in that block. After finishing the block, decoding continues with the next block count, until a 0 count terminates the sequence.

Here's the text from the [Avro Specification](https://avro.apache.org/docs/1.11.1/specification/#maps) regarding maps with negative block counts:

> If a block’s count is negative, its absolute value is used, and the count is followed immediately by a long block size indicating the number of bytes in the block. This block size permits fast skipping through data, e.g., when projecting a record to a subset of its fields.

klion26 commented on arrow-avro/src/reader/record.rs:
```diff
@@ -290,3 +401,84 @@ fn flush_primitive<T: ArrowPrimitiveType>(
+    }
+
+    fn avro_from_codec(codec: Codec) -> AvroDataType {
+        AvroDataType::new(codec, None, Default::default())
```
Not sure if we need to test the `nullability` to `Some(x)`

jecsand838 replied:
@klion26 I was thinking it would be best to handle nullability in a dedicated PR after the basic support for each type is merged in.

jecsand838 commented:
@klion26 I pushed changes just now that:

* Fixed the linter errors
* Made `read_blockwise_items` more readable
* Changed `let val_field = Field::new("value", val_dt, true)` to  `let val_field = Field::new("value", val_dt, value_type.nullability.is_some())`

Let me know what you think!

klion26 reviewed (approved):
@jecsand838 thanks for the change, LGTM

alamb commented on arrow-avro/src/codec.rs:
```diff
@@ -162,6 +175,8 @@ pub enum Codec {
     List(Arc<AvroDataType>),
     /// Represents Avro record type, maps to Arrow's Struct data type
     Struct(Arc<[AvroField]>),
+    /// Represents Avro map type, maps to Arrow's Map data type
```
Since this is a `pub` enum and it is not marked as `non exhaustive` I think this is technically a breaking API change and will need to wait for the next major release:
- #7395

alamb reviewed (approved):
Thanks you -- looks good to me @jecsand838 

Thank you @klion26  for the review

tustvold commented:
I've not reviewed the code, but as this is not actually public (none of the avro crate is) it doesn't need to wait for a breaking release

alamb commented:
> I've not reviewed the code, but as this is not actually public (none of the avro crate is) it doesn't need to wait for a breaking release

In that case, let's get it merged!

alamb commented:
Thanks @jecsand838  -- it is exciting to see progress here

jecsand838 commented:
@alamb @tustvold Absolutely! We'll go ahead and get some more PRs up this week then! Thank you all much for getting this in.
